closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1040 @@
+                            for (String attribute : attributeArray) {
+                                String[] in = attribute.split("=", 2);
+                                if (in.length != 2) {
+                                    printStackTraceAndExit(
+                                            "Constraint attributes format is incorrect: " + providerInfo);
+                                }
+                            }
+                        }
+
+                        Constraint constraint = new Constraint(inType, inAlgorithm, inAttributes);
+
+                        if (debug != null) {
+                            debug.println("Loading constraints for security provider: " + providerName
+                                    + " with constraints type: " + constraint.type
+                                    + " algorithm: " + constraint.algorithm
+                                    + " attributes: " + constraint.attributes);
+                        }
+                        constraints[cNum] = constraint;
+                        cNum++;
+                    }
+                    providerConstraints.put(providerName, constraints);
+                    if (debug != null) {
+                        debug.println("Loaded constraints for security provider: " + providerName);
+                    }
+                } else {
+                    printStackTraceAndExit("Constraint format is incorrect: " + providerInfo);
+                }
+            }
+        }
+
+        /**
+         * Check if the Service is allowed in restricted security mode.
+         *
+         * @param service the Service to check
+         * @return true if the Service is allowed
+         */
+        boolean isRestrictedServiceAllowed(Service service) {
+            String providerName = service.getProvider().getName();
+
+            // Provider with argument, remove argument.
+            // e.g. SunPKCS11-NSS-FIPS, remove argument -NSS-FIPS.
+            int pos = providerName.indexOf('-');
+            providerName = (pos < 0) ? providerName : providerName.substring(0, pos);
+
+            Constraint[] constraints = providerConstraints.get(providerName);
+
+            if ((constraints == null) || (constraints.length == 0)) {
+                // Allow this provider with no constraints.
+                return true;
+            }
+
+            // Check the constraints of this provider.
+            String type = service.getType();
+            String algorithm = service.getAlgorithm();
+
+            for (Constraint constraint : constraints) {
+                String cType = constraint.type;
+                String cAlgorithm = constraint.algorithm;
+                String cAttribute = constraint.attributes;
+
+                boolean cTypePut = isAsterisk(cType)
+                        || type.equals(cType);
+                boolean cAlgorithmPut = isAsterisk(cAlgorithm)
+                        || algorithm.equals(cAlgorithm);
```

Comment:
Algorithms should be case-insensitive, this should use `equalsIgnoreCase()` instead of `equals()`.

This can be simplified, avoiding the need for *Put variables that are tested repeatedly:
```java
                if (!isAsterisk(cType) && !type.equals(cType)) {
                    // The constraint doesn't apply to the service type.
                    continue;
                }
                if (!isAsterisk(cAlgorithm) && !algorithm.equalsIgnoreCase(cAlgorithm)) {
                    // The constraint doesn't apply to the service algorith.
                    continue;
                }
```